Measure the new kernel as an EFI binary

aplanas opened the pull request:
Also update the version to 0.5.1

aplanas commented:
cc: @okirch

okirch commented on src/efi-application.c:
```diff
@@ -288,10 +290,20 @@ __tpm_event_efi_bsa_rehash(const tpm_event_t *ev, const tpm_parsed_event_t *pars
 		return __efi_application_rehash_pesign(ctx, evspec->efi_partition, evspec->efi_application);
 
-	return __efi_application_rehash_direct(parsed, ctx);
+	return __efi_application_rehash_direct(&evspec_clone, ctx);
```
This looks like it will fail if we're hashing something other than a kernel. If sdb_is_kernel returns false, evspec_clone remains uninitialized. So better call __efi_application_rehash_direct(evspec, ctx)

aplanas replied:
omg ... yes. Fixed

okirch commented on src/efi-application.c:
```diff
@@ -288,10 +290,20 @@ __tpm_event_efi_bsa_rehash(const tpm_event_t *ev, const tpm_parsed_event_t *pars
+	if (sdb_is_kernel(evspec->efi_application)) {
+		new_application = sdb_get_next_kernel();
+		if (new_application) {
+			memcpy(&evspec_clone, evspec, sizeof(evspec_clone));
```
I'd use evspec_clone = *evspec; instead. Does the same thing but you get type checking for free.

aplanas replied:
Good idea. Done!

okirch commented on src/sd-boot.c:
```diff
@@ -294,25 +297,20 @@ exists_efi_dir(const char *path)
-bool
-sdb_get_entry_list(sdb_entry_list_t *result)
+static const char *
+get_token_id()
```
BTW why not get_token_id(void) ?

aplanas replied:
Yes. Should be fixed.

okirch commented on src/sd-boot.c:
```diff
@@ -324,6 +322,22 @@ sdb_get_entry_list(sdb_entry_list_t *result)
+	struct dirent *dir;
+	char *path = "/boot/efi/loader/entries";
+
+	memset(result, 0, sizeof(*result));
```
It's just a matter of taste, but I usually like the caller to initialize a variable before it gets passed into some function like sdb_get_entry_list.

The slightly bigger issue is the risk of stack buffer overruns if the file has more than SD_MAX_ENTRIES matching entries. Somewhere inside the while (readdir()) loop you should check num_entries and error out when it exceeds the limit.

aplanas replied:
> caller to initialize a variable
Done.

> bigger issue is the risk of stack buffer overruns
Should be done

okirch commented on src/sd-boot.h:
```diff
@@ -31,6 +31,7 @@ typedef struct sdb_entry_data {
 	char	machine_id[SDB_LINE_MAX];
 	char	version[SDB_LINE_MAX];
 	char	options[SDB_LINE_MAX];
+	char	linux_[SDB_LINE_MAX];
```
I assume the trailing underscore is there because linux is a cpp macro, but it looks odd :-)
Please consider using some other identifier like "kernel" or "image"

okirch commented on src/sd-boot.h:
```diff
@@ -40,5 +41,7 @@ typedef struct sdb_entry_list {
 
 extern bool			sdb_get_entry_list(sdb_entry_list_t *result);
+extern bool			sdb_is_kernel(const char *application);
+extern const char *		sdb_get_next_kernel();
```
Again, I'd use (void) in this declaration.

aplanas replied:
Done

aplanas commented:
cc @okirch 

I added the support for EV_EVENT_TAG from the kernel (PRC#9).  It will emit two, one for the command line and another for the initrd measurement.

Also updated the version to 0.5.2

aplanas commented:
@okirch can be possible that for each upgrade of the version file we have a new tag associated? This will help when updating the package.

okirch commented:
Merged, thank you.

aplanas commented:
> FWIW, I also created tags for 0.5.* and pushed them

Thanks! That will help a lot.

okirch commented:
I just pushed 2 patches to a new branch called "future". I'd appreciate if you could check them out.
One is just a small cosmetic one that creates a new helper function for unix2dos path conversion. Your open-coded version didn't 0-terminate the string when the initrd path was too long, so while I was fixing it I moved it out of eventlog.c into its own little function

The other is a somewhat bigger design change. I really think it needs to be a user decision whether to hash a new kernel and initrd, and if so, which one. (In the longer term, I want to be able to run pcr-oracle on a central management host and be able to create signed policies for an organization's clients.)
So the second patch introduces an option called --next-kernel, which is supposed to fill in various fields in the rehash_ctx struct. Which means that the user has control over which kernel we rehash; plus we do this before processing the event log. The idea is that "--next-kernel auto" will do what your code is doing now, while "--next-kernel <someid>" will use the sdb entry identified by <someid>. I'm sure there will be more creative variations (maybe for rollback it would be useful to support "previous").

Please take a look at the patch and let me know what you think. I haven't implemented all the details, just the interface.

I'll be gone until Sunday, so please do not expect anything from me before then :-)

aplanas commented:
> I just pushed 2 patches to a new branch called "future". I'd appreciate if you could check them out. One is just a small cosmetic one that creates a new helper function for unix2dos path conversion.

Thanks for this. I commented there about this.

> The other is a somewhat bigger design change. I really think it needs to be a user decision whether to hash a new kernel and initrd, and if so, which one.

I very much agree. For me one problem is that we need somehow create multiple predictions stored in the JSON file. The document allows that, but pcr-oracle if using functions that have an internal state (static variable), making it hard to reuse in iterations.

The command like will help in this direction, as can be called by the script with the expected combinations of parameters.

> (In the longer term, I want to be able to run pcr-oracle on a central management host and be able to create signed policies for an organization's clients.) So the second patch introduces an option called --next-kernel, which is supposed to fill in various fields in the rehash_ctx struct. Which means that the user has control over which kernel we rehash; plus we do this before processing the event log. The idea is that "--next-kernel auto" will do what your code is doing now, while "--next-kernel someid" will use the sdb entry identified by someid. I'm sure there will be more creative variations (maybe for rollback it would be useful to support "previous").

BTW, there is a nice tool in the next systemd (v255) that is covering much of what pcr-oracle is doing wrt PCR value prediction: https://www.freedesktop.org/software/systemd/man/latest/systemd.pcrlock.html

I think that we can learn something from this approach.

> Please take a look at the patch and let me know what you think. I haven't implemented all the details, just the interface.
> 
> I'll be gone until Sunday, so please do not expect anything from me before then :-)
